### ISSUE TITLE
Duplicate chain command

### ISSUE THREAD

**Issue by Piattica** [zynthian/zynthian-issue-tracking#1040]
**Is your feature request related to a problem? Please describe.**
When I am preparing snapshots for a song it would be useful to have a duplicate chain command in menu to make faster the preparation

**Describe the solution you'd like**
in MIXER screen, go over a chain and with Select to have a **duplicate chain** command

**Comment by riban-bw** [comment]
This has been discussed at Zynthclub but I want to know more about the workflow to avoid implementation of something that doesn't provide sufficient benefit.

Will you please give some examples of why a duplicate function would be beneficial? This might include an explaination of how the duplicate chain is subsequently modified and why.

We should consider how much is duplicated. Do we create a full chain with all processors? Do we copy the processor parameters? Do we copy the presets? How do we handle processors that can not be duplicated, e.g. Pianoteq is a single instance engine?

**Comment by Piattica** [comment]
as example I prepare a split with 1st chain a string section split A0-B3, 2nd chain a flute C4-G8 and 3rd chain as the 2nd but with a synth and 4th chain as the 2nd but with a pad
so for me is convenient to duplicate the 2nd chain to build a snapshot

**Comment by riban-bw** [comment]
Ha! There are parameters I hadn't mentioned... split points would also want to be duplicated. My _engineer head_ tells me that we could have a flexible, configurable implementation that allows duplicating some or all of the parameters and copying some or all parameters between chains... but my _project manager head_ says that we need something that is simple to comprehend and use and we should avoid complexities that some (most) users may never want or need.

Even in that thought we see another possible feature: copy. And in there the question - do we want or need both?

Maybe combining so that you can copy (a selection of parameters) from one chain and apply them to another, including creating a new chain based on those parameters.

We also need to consider MIDI channel. When you create a MIDI / Instrument chain you are asked what MIDI channel to use. If you were duplicating a chain, how many questions do we want to ask?

Maybe a copy feature that pops up a checkbox list of options describing what to copy (with a sane set of defaults or maybe the previously selected options) and then a list of chains, including "New chain" as the target to copy to. The list could be multi-select allowing targetting the copy at many chains which would allow you to take the key-split set up on chain 1 and apply it to chains 2, 3, 4, etc.

**Comment by Piattica** [comment]
> Maybe a copy feature that pops up a checkbox list of options describing what to copy (with a sane set of defaults or maybe the previously selected options) and then a list of chains, including "New chain" as the target to copy to.

It would be great but listen also the other users

**Comment by jofemodo** [comment]
So, it seems you only want to duplicate the split points, not the full chain.
Perhaps other user wants to duplicate audio routing, for instance. Etc. The possibilities are endless, so the UI complexity needed to manage all these possibilities. You would end doing the splitting by hand because the "duplication menu" has too many options and it's confusing for you ;-DDD
I don't see a huge advantage in implementing something lie this: the effort for setting the split points by hand is low and currently it's a single-user case.
In the other hand, i see more interesting to save/load independent chains, what could include a kind of "chain-templates". This could be an interesting approach and it would also solve the use-case, right?

Regards

**Comment by riban-bw** [comment]
The concept of copying a chain is something that has been mentioned before and it seems like a natural workflow but the complexity involved in selecting elements to copy / paste seems high. Having the ability to save "snapshots" with single chairs and/or merging part/all of a snapshot does seem useful and, although it may seem like an extra step, it isn't really much different. It has the additional benefit of providing a template workflow which may actually be a better workflow for use cases like this.

Maybe the "New Chain"  menu could offer "From template..." option which could show a list of chains that exist in snapshots. (Maybe all snapshots or maybe a filtered list or maybe all with an extra "Templates" folder.

We _could_ offer some filters on what gets loaded, like audio routing, split points, etc. but that probably only makes sense if we are able to merge into existing chains which, as @jofemodo says, could be confusing.

We should consider the workflows to determine the best solution.

**Comment by Piattica** [comment]
I agree: copying a chain menu voice would be useful but it is complex depending to esult you need (instrument? intrument + effects? note range etc.). The risk is to have a too big sub-menu that at the end nobody will use.
The way of  template snapshots is the easier: one it will be useful to provide a template snapshots folder with the more common templates well set. 
I am already use this way.

**Comment by riban-bw** [comment]
I have implemented a new feature in vangelis (testing) which allows you to export a single chain as a snapshot. This can allow creation of single chain templates. It is accessed from the chain option menu. There is another feature to import or merge snapshots which will add any chains within a snapshot to those already configured in the zynthian. This is accessed from the snapshot menu.

Although this is more cumbersome than copying a snapshot, you can do so with this mechanism and it may prove sufficient.
